runtime/src/epoch_stakes.rs

```diff
@@ -134,13 +138,58 @@ impl EpochStakes {
 #[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
 pub enum VersionedEpochStakes {
     Current {
-        stakes: Stakes<Stake>,
+        stakes: StakesSerdeWrapper,
         total_stake: u64,
         node_id_to_vote_accounts: Arc<NodeIdToVoteAccounts>,
         epoch_authorized_voters: Arc<EpochAuthorizedVoters>,
     },
 }
 
+/// Wrapper struct with custom serialization to support serializing
+/// `Stakes<StakeAccount>` as `Stakes<Stake>` without doing a full deep clone of
+/// the stake data. Serialization works by building a `Stakes<&Stake>` map which
+/// borrows `&Stake` from `StakeAccount` entries in `Stakes<StakeAccount>`. Note
+/// that `Stakes<&Stake>` still copies `Pubkey` keys so the `Stakes<&Stake>`
+/// data structure still allocates a fair amount of memory but the memory only
+/// remains allocated during serialization.
+#[cfg_attr(feature = "frozen-abi", derive(AbiExample, AbiEnumVisitor))]
+#[derive(Debug, Clone, PartialEq)]
+pub enum StakesSerdeWrapper {
+    Stake(Stakes<Stake>),
+    Account(Stakes<StakeAccount<Delegation>>),
+}
+
+impl From<StakesSerdeWrapper> for StakesEnum {
+    fn from(stakes: StakesSerdeWrapper) -> Self {
+        match stakes {
+            StakesSerdeWrapper::Stake(stakes) => Self::Stakes(stakes),
+            StakesSerdeWrapper::Account(stakes) => Self::Accounts(stakes),
+        }
+    }
+}
+
+impl Serialize for StakesSerdeWrapper {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match self {
+            Self::Stake(stakes) => stakes.serialize(serializer),
+            Self::Account(stakes) => Stakes::<&Stake>::from(stakes).serialize(serializer),
```

Comment:
The conversion from `Stakes<StakeAccount>` to `Stakes<Stake>` is now done here during serialization and does so in a more memory efficient manner by converting to `Stakes<&Stake>` rather than `Stakes<Stake>`

Reply:
I think a side-effect of this is that the deserialized value of serialized `StakesSerdeWrapper` is no longer equal to the original value. iow serialize -> deserialize does not round trip.
 
Don't we need a manual implementation for `PartialEq` as well?

Reply:
Yeah we do, thanks. Will add that